openxr/examples/vulkan.rs

```diff
@@ -0,0 +1,649 @@
+        let fences = (0..PIPELINE_DEPTH)
+            .map(|_| {
+                vk_device
+                    .create_fence(
+                        &vk::FenceCreateInfo::builder().flags(vk::FenceCreateFlags::SIGNALED),
+                        None,
+                    )
+                    .unwrap()
+            })
+            .collect::<Vec<_>>();
+
+        // Main loop
+        let mut swapchain = None;
+        let mut event_storage = xr::EventDataBuffer::new();
+        let mut session_running = false;
+        // Index of the current frame, wrapped by PIPELINE_DEPTH. Not to be confused with the
+        // swapchain image index.
+        let mut frame = 0;
+        'main_loop: loop {
+            if !running.load(Ordering::Relaxed) {
+                println!("requesting exit");
+                // The OpenXR runtime may want to perform a smooth transition between scenes, so we
+                // can't necessarily exit instantly. Instead, we must notify the runtime of our
+                // intent and wait for it to tell us when we're actually done.
+                match session.request_exit() {
+                    Ok(()) => {}
+                    Err(xr::sys::Result::ERROR_SESSION_NOT_RUNNING) => break,
+                    Err(e) => panic!("{}", e),
+                }
+            }
+
+            while let Some(event) = instance.poll_event(&mut event_storage).unwrap() {
+                use xr::Event::*;
+                match event {
+                    SessionStateChanged(e) => {
+                        println!("entered state {:?}", e.state());
+                        match e.state() {
+                            xr::SessionState::READY => {
+                                session
+                                    .begin(xr::ViewConfigurationType::PRIMARY_STEREO)
+                                    .unwrap();
+                                session_running = true;
+                            }
+                            xr::SessionState::STOPPING => {
+                                session.end().unwrap();
+                                session_running = false;
+                            }
+                            xr::SessionState::EXITING | xr::SessionState::LOSS_PENDING => {
+                                break 'main_loop;
+                            }
+                            _ => {}
+                        }
+                    }
+                    InstanceLossPending(_) => {
+                        break 'main_loop;
+                    }
+                    EventsLost(e) => {
+                        println!("lost {} events", e.lost_event_count());
+                    }
+                    _ => {}
+                }
+            }
+
+            if !session_running {
+                // Don't grind up the CPU
+                std::thread::sleep(Duration::from_millis(100));
+                continue;
+            }
+
+            let state = frame_wait.wait().unwrap();
```

Comment:
Maybe using an xr prefix could make this clearer. This confused me for a hot second before I realised it's not a vulkan helper. Since the vulkan stuff is mostly prefixed, maybe prefixing the xr stuff would be more consistent.